octobot_trading/personal_data/portfolios/history/historical_portfolio_value_manager.py

```diff
@@ -0,0 +1,241 @@
+                commons_enums.TimeFramesMinutes[time_frame] * commons_constants.MINUTE_TO_SECONDS
+        ))
+
+    def _get_relevant_timestamps(self, timestamp, currencies, time_frames, force_update):
+        relevant_timestamps = []
+        current_time = self.portfolio_manager.exchange_manager.exchange.get_exchange_current_time()
+        for time_frame in time_frames:
+            # allowed times are [local time frame t0: local time frame t0 + allowed lag]
+            time_frame_allowed_window_start = self.convert_to_historical_timestamp(current_time, time_frame)
+            if self._should_update_timestamp(currencies, time_frame_allowed_window_start, force_update):
+                # allow time window if time_frame_allowed_window_start not in self.historical_portfolio_value
+                # or when force_update
+                time_frame_seconds = commons_enums.TimeFramesMinutes[time_frame] * commons_constants.MINUTE_TO_SECONDS
+                time_frame_allowed_window_end = time_frame_allowed_window_start + \
+                    max(time_frame_seconds * (1 + self.TIME_FRAME_RELEVANCY_TIME_RATIO),
+                        time_frame_seconds + self.MIN_TIME_FRAME_RELEVANCY_SECONDS)
+                if time_frame_allowed_window_start <= timestamp <= time_frame_allowed_window_end:
+                    relevant_timestamps.append(time_frame_allowed_window_start)
+        return relevant_timestamps
+
+    def _should_update_timestamp(self, currencies, time_frame_allowed_window_start, force_update):
+        if force_update:
+            return True
+        try:
+            for currency in currencies:
+                self.get_historical_value(time_frame_allowed_window_start).get(currency)
+        except KeyError:
+            return True
+        return False
+
+    def _get_value_in_currency(self, historical_value, currency):
+        if currency in historical_value:
+            return historical_value.get(currency)
+        return self._convert_historical_value(historical_value, currency)
+
+    def _convert_historical_value(self, historical_value, target_currency):
+        # TODO try to get a more accurate historical value into target_currency currency using price history
+        # last chance: try to get any usable value from portfolio value holder (not accurate since used the intermediary
+        # asset might also have changed in price since the time it was recorded)
+        for currency in historical_value.get_currencies():
+            for pair, price in self.portfolio_manager.portfolio_value_holder.last_prices_by_trading_pair.items():
+                base, quote = symbol_util.split_symbol(pair)
+                if base == currency:
+                    return historical_value.get(currency) * price
+                if quote == currency:
+                    return historical_value.get(currency) / price
+        raise errors.MissingPriceDataError(f"no price data to evaluate {historical_value} on {target_currency}")
```

Comment:
Don't we have a similar code somewhere else? In portfolio value holder I think

Reply:
We do in PortfolioValueHolder with a system to register missing pairs in channels etc, here we just want to read prices if available, I think we should keep it simple like this

Reply:
We can't just write an util method to prevent duplicating this code?